### PR TITLE
[Proof of Concept] Azure Monitor Exporter Integration Tests

### PR DESCRIPTION
**Pull request by TimothyMothra**
This is a proof of concept building an Integration Test project for the AzureMonitor Exporter.

## Proposal
I propose that we expose just enough of the internals of the Exporter to offer deterministic testing as a feature.
I would like to see this as a feature fully documented in our public docs. 
Providing apis for testing our scenarios should be satisfactory for customer's scenarios as well.

## Changes
- New project `IntegrationTests`. This is where tests are executed.
- New project `IntegrationTests.WebApp`. This is a web app that can be dynamically instrumented and configured from a test.
- `AzureMonitorTransmitter` was modified to accept a Func from `AzureMonitorExporterOptions` that can override the behavior of sending TelemetryItems to the Ingestion Service.

### API Changes
- Added `AzureMonitorExporterOptions.OnTrackAsync`.
The very final action the Exporter/Transmitter take is to pass `IEnumerable<TelemetryItem>` to the `ServiceRestClient`. I've exposed this as a Func so the behavior can be overridden in a test application. 
```
        /// <summary>
        /// This property can be used to change the behavior of the <see cref="AzureMonitorTraceExporter"/>.
        /// This will divert any <see cref="TelemetryItem"/> from the Ingestion Service.
        /// This is provided for testing.
        /// </summary>
        public Func<IEnumerable<TelemetryItem>, CancellationToken, Task<Response<TrackResponse>>> OnTrackAsync { get; set; } = null;
```

-  Modified the `AzureMonitorExporterHelperExtensions.AddAzureMonitorTraceExporter` method to optionally expose the `ActivityProcessor`.
There are no public APIs to get this from OpenTelemetry's `TracerProviderBuilder`.
**This is not final.** I'm using this to call `ActivityProcessor.ForceFlush` at the end of a test.

## TODO
- How to mock the `Azure.Response<TrackResponse>` object from the `ServiceRestClient`? This would be useful for mocking behavior from the Ingestion Service.

### CODE REVIEW

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/tests/OpenTelemetry.Exporter.AzureMonitor.IntegrationTests/OpenTelemetryTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            // Act
+            var response = await client.GetAsync(request);
+
+            // Assert
+            response.EnsureSuccessStatusCode();
+
+            // TODO: HOW TO REMOVE THE WAIT? CAN THE EXPORTER BE FLUSHED?
+            Task.Delay(5000).Wait();
+
+            /// TODO: I'M TRYING TO CALL SHUTDOWN TO GUARENTEE THAT THE EXPORTER IS FLUSHED WITHOUT NEEDING A WAIT.
```
Comment (reyang) [inline sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/tests/OpenTelemetry.Exporter.AzureMonitor.IntegrationTests/OpenTelemetryTests.cs:39]: Refer to the spec:
* https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#shutdown
* https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#shutdown-1

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs [hunk 2]
```diff
@@ -68,17 +73,33 @@ internal async ValueTask<int> AddBatchActivityAsync(Batch<Activity> batchActivit
...
-            return response.Value.ItemsAccepted.GetValueOrDefault();
+            return response?.Value.ItemsAccepted.GetValueOrDefault() ?? default(int);
         }
 
+        /// <summary>
+        /// In a normal configuration, this is expected to invoke the serviceRestClient and transmit telemetry to the ingestion service.
+        /// This property can be changed to override the behavior of transmitting telemetry.
+        /// This can be used for Unit Testing to capture telemetry that would have been transmitted.
+        /// </summary>
+        internal Func<IEnumerable<TelemetryItem>, CancellationToken, Task<Response<TrackResponse>>> OnTrackAsync { private get; set; }
```
Comment (reyang) [inline sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: Why do we need this at all?

Comment (TimothyMothra) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: I'm open to alternate approaches. 
The very last action the Exporter/Transmitter execute is to send telemetry to the ServiceRestClient.
I want the unit tests to be able change this behavior and say, send telemetry to me instead.

- The goal here is to allow telemetry to pass fully through the Exporter. 
- I want the actual TelemetryItem object so we're not trying to capture http requests or serialize Json within a test.

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/tests/OpenTelemetry.Exporter.AzureMonitor.IntegrationTests/OpenTelemetryTests.cs (hunk 1 above)
Comment (TimothyMothra) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/tests/OpenTelemetry.Exporter.AzureMonitor.IntegrationTests/OpenTelemetryTests.cs:39]: woohoo! looks like ForceFlush on the ActivityProcessor does the trick! :)

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs (hunk 2 above)
Comment (reyang) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: Consider mock the underlying ServiceRestClient.

Comment (TimothyMothra) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: I thought of that but the current `ServiceRestClient` is internal. 
I could instead introduce a new interface; `IServiceRestClient` and keep the current implementation internal.

Would you approve of that?

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorExporterHelperExtensions.cs [hunk 3]
```diff
@@ -18,6 +18,18 @@ public static class AzureMonitorExporterHelperExtensions
...
+        }
+
+        /// <summary>
+        /// Registers an Azure Monitor trace exporter that will receive <see cref="System.Diagnostics.Activity"/> instances.
+        /// </summary>
+        /// <param name="builder"><see cref="TracerProviderBuilder"/> builder to use.</param>
+        /// <param name="configure">Exporter configuration options.</param>
+        /// <param name="processor">WORK IN PROGRESS.</param>
+        /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+        public static TracerProviderBuilder AddAzureMonitorTraceExporter(this TracerProviderBuilder builder, out ActivityProcessor processor, Action<AzureMonitorExporterOptions> configure)
```
Comment (reyang) [inline sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorExporterHelperExtensions.cs:32]: Why do we want this in the function signature?

Comment (TimothyMothra) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorExporterHelperExtensions.cs:32]: To be honest, I don't :)
But i couldn't find any public APIs get a handle on the ActivityProcessor.
I can get a handle on the `TracerProviderBuilder`, but it's processor is private.

This part is a work in progress, proving that calling ForceFlush works.

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs (hunk 2 above)
Comment (reyang) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: > I thought of that but the current `ServiceRestClient` is internal.

What's the blocking issue with "the current `ServiceRestClient` is internal"?

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorExporterHelperExtensions.cs (hunk 3 above)
Comment (reyang) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorExporterHelperExtensions.cs:32]: var exporter = new AzureMonitorTraceExporter(...);
var processor = new BatchExportActivityProcessor(exporter);

builder.AddProcessor(processor).Build();

// do something

var result = processor.ForceFlush(timeout);

// do something with result

File: sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs (hunk 2 above)
Comment (TimothyMothra) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: I want to be able to provide a Public API for unit testing.
My IntegrationTest app currently has no access to internal classes, so I won't be able to mock the ServiceRestClient using a tool like Moq. 
Maybe I'm not following if you have something else in mind.

Comment (reyang) [reply sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs:92]: This is what's in my mind:
* We don't want to expose public API "just for our own testing purpose".
* Test cases should have access to internal members of the target.

**Pull request closed without merging**